hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/ReadHdfsNewApiP.java

```diff
@@ -0,0 +1,293 @@
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * See {@link HdfsSources#hdfs}.
+ */
+public final class ReadHdfsNewApiP<K, V, R> extends AbstractProcessor {
+
+    private static final Class<?>[] EMPTY_ARRAY = new Class[0];
+
+    private final Configuration configuration;
+    private final InputFormat inputFormat;
+    private final Traverser<R> trav;
+    private final BiFunctionEx<K, V, R> projectionFn;
+    private final BetterByteArrayOutputStream cloneBuffer = new BetterByteArrayOutputStream();
+    private final DataOutputStream cloneBufferDataOutput = new DataOutputStream(cloneBuffer);
+
+    private InternalSerializationService serializationService;
+
+    private ReadHdfsNewApiP(
+            @Nonnull Configuration configuration,
+            @Nonnull InputFormat inputFormat,
+            @Nonnull List<InputSplit> splits,
+            @Nonnull BiFunctionEx<K, V, R> projectionFn
+    ) {
+        this.configuration = configuration;
+        this.inputFormat = inputFormat;
+        this.trav = traverseIterable(splits)
+                .flatMap(this::traverseSplit);
+        this.projectionFn = projectionFn;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        HazelcastInstanceImpl instance = (HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance();
+        serializationService = instance.getSerializationService();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(trav);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Traverser<R> traverseSplit(InputSplit split) {
+        RecordReader<K, V> reader;
+        try {
+            TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(configuration, new TaskAttemptID());
+            reader = inputFormat.createRecordReader(split, attemptContext);
+            reader.initialize(split, attemptContext);
+        } catch (IOException | InterruptedException e) {
+            throw sneakyThrow(e);
+        }
+
+        return () -> {
+            try {
+                while (reader.nextKeyValue()) {
+                    // we clone the key/value if configured so because some of the
+                    // record-readers return the same object for `reader.getCurrentKey()`
+                    // and `reader.getCurrentValue()` which is mutated for each `reader.nextKeyValue()`
+                    K key = copyIfConfigured(reader.getCurrentKey());
+                    V value = copyIfConfigured(reader.getCurrentValue());
+                    R projectedRecord = projectionFn.apply(key, value);
+                    if (projectedRecord != null) {
+                        return projectedRecord;
+                    }
+                }
+                reader.close();
```

Comment:
this reader won't be closed if exception

Reply:
Also won't be closed when the job is cancelled.

Reply:
extracted to a field, now closed in `close` method